guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+  });
+});
+```
+
+💡 Now say the global you want to change is not a function:
+
+🚫
+
+``` typescript
+describe("interpretMethodData", () => {
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    // This change will apply to every other test that's run after this one
+    global.ethereumProvider = new HttpProvider(
+      'https://sepolia.infura.io/v3/abc123',
+    );
+    expect(interpretMethodData('0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    // Oops! This will still hit Sepolia
+    expect(interpretMethodData('0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+[Instead of using hooks](#avoid-before-each-and-after-each), one idea is to create a function wrapping your test which captures the current value of the global beforehand and restores the global to this value afterward:
+
+1️⃣
+
+``` typescript
+async function withEthereumProvider(ethereumProvider: Provider, test: () => void | Promise<void>) {
+  const originalEthereumProvider = global.ethereumProvider;
+  global.ethereumProvider = ethereumProvider;
+  await fn();
+  global.ethereumProvider = originalEthereumProvider;
+}
+
+describe("interpretMethodData", () => {
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    const sepolia = new HttpProvider('https://sepolia.infura.io/v3/abc123');
+    withEthereumProvider(sepolia, () => {
+      expect(interpretMethodData('0x3fbac0ab')).toStrictEqual({
+        name: 'Set Approval For All'
+      });
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    // Now this test will "just work"
+    expect(interpretMethodData('0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+Even better, however, is to make use of [dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) to eliminate the need for a global. This allows you to create a fake version of the value in question within your test:
+
+2️⃣
+
+``` typescript
+describe("interpretMethodData", () => {
+  it("returns the signature for setApprovalForAll on Sepolia", () => {
+    const provider = createFakeProvider({ network: "sepolia" });
+    expect(interpretMethodData(provider, '0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+
+  it("returns the signature for setApprovalForAll on Mainnet", () => {
+    const provider = createFakeProvider({ network: "mainnet" });
+    expect(interpretMethodData(provider, '0x3fbac0ab')).toStrictEqual({
+      name: 'Set Approval For All'
+    });
+  });
+});
+```
+
+### Shared variables
+
+If a variable used inside of a test is declared outside of the test, the value that the variable holds — and thus changes to that variable made inside the test — will remain even after the test ends.
```

Comment:
Nit: I found the wording here to be a little confusing. I've attempted to re-word it while conveying the same information, do you think this sounds OK?
```suggestion
Variables declared outside of tests are not reset automatically between tests. Changes made to these variables in tests can affect later tests, breaking test isolation.
```

Reply:
Good suggestion. I didn't see this before I revised this whole section, but I've applied the suggestion here: [48d3c7a](https://github.com/MetaMask/contributor-docs/pull/38/commits/48d3c7a481a227f9527ca03fc80eac0a6c6e404c)